tests/modeltests/tst_callgraphgenerator.cpp

```diff
@@ -0,0 +1,139 @@
+#include "callgraphgenerator.h"
+
+#include <QDebug>
+#include <QObject>
+#include <QSignalSpy>
+#include <QTemporaryFile>
+#include <QTest>
+
+#include "../../src/parsers/perf/perfparser.h"
+#include "../../src/perfrecord.h"
+#include "../testutils.h"
+#include "data.h"
+
+QString findExe(const QString& name)
+{
+    QFileInfo exe(QCoreApplication::applicationDirPath() + QLatin1String("/../test-clients/%1/%1").arg(name));
+    qDebug() << exe;
+    VERIFY_OR_THROW(exe.exists() && exe.isExecutable());
+    return exe.canonicalFilePath();
+}
+
+class TestCallgraphGenerator : public QObject
+{
+    Q_OBJECT
+private slots:
+    void initTestCase()
+    {
+        qSetGlobalQHashSeed(0);
+
+        qRegisterMetaType<Data::BottomUpResults>("Data::BottomUpResults");
+        qRegisterMetaType<Data::CallerCalleeResults>("Data::CallerCalleeResults");
+        qRegisterMetaType<Data::EventResults>("Data::EventResults");
+
+        const QStringList perfOptions = {"--call-graph", "dwarf"};
+        QStringList exeOptions;
+
+        const QString exePath = findExe("callgraph");
+        m_file.open();
+
+        perfRecord(perfOptions, exePath, exeOptions, m_file.fileName());
+    }
+
+    void testParent()
+    {
+        CallerCalleeModel model;
+        const auto results = callerCalleeResults(m_file.fileName());
+        model.setResults(results);
+
+        QVERIFY(callerCalleeResults(m_file.fileName()).entries.size() > 0);
+
+        auto key = Data::Symbol();
+        for (const auto& entry : results.entries.keys()) {
```

Comment:
please eradicate this antipattern from your brain, it's really really really bad - use a proper iterator instead